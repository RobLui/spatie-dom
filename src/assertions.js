export function hasTag(el, tag) {
    return el.tagName.toLowerCase() === tag.toLowerCase();
}

export function matches(el, selector) {
    if (el.matches) {
        return el.matches(selector);
    }

    if (el.msMatchesSelector) {
        return el.msMatchesSelector(selector);
    }

    throw new Error('`Element.matches` is not supported in this browser');
}
