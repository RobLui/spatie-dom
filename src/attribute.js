export function attribute(name, element, fallback) {
    return element.getAttribute(name) || fallback;
}