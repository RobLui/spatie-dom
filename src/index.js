export * from './dom/props';
export * from './dom/query';
export * from './dom/when';
