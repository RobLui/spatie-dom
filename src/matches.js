export function matches(element, selector) {
    if (element.matches) {
        return element.matches(selector);
    }

    if (element.msMatchesSelector) {
        return element.msMatchesSelector(selector);
    }

    throw new Error('`Element.matches` is not supported in this browser');
}