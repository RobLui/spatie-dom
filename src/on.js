export function on(event, subject, handler) {
    subject.addEventListener(event, handler);
}