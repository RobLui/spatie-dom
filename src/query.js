export function query(selector, scope = document) {
    return scope.querySelector(selector);
}

export function queryAll(selector, scope = document) {
    return [...scope.querySelectorAll(selector)];
}

export const $ = query;
export const $$ = queryAll;