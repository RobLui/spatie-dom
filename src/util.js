export function camelCase(string) {
    return string.replace(
        /(-|_)([a-z])/g, (result, delimeter, char) => char.toUpperCase()
    );
}

export function kebabCase(string) {
    return string.replace(
        /([A-Z])/g, char => '-' + char.toLowerCase()
    );
}
