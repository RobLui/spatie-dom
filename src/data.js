import { attribute } from './attribute';

export function data(name, element, fallback = '') {
    return attribute(`data-${name}`, element, fallback);
}