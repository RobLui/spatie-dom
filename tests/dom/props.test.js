import { query, prop, props } from '../../src';

describe('prop', () => {

    it('can retrieve a text prop from an element', () => {
        document.body.innerHTML = `
            <div id="el" my-prop="foo"></div>
        `;

        const myProp = prop(query('#el'), 'myProp');

        expect(myProp).toBe('foo');
    });

    it('can retrieve an evaluated json prop from an element', () => {
        document.body.innerHTML = `
            <div id="el" :my-prop='{ "foo": "bar" }'></div>
        `;

        const myProp = prop(query('#el'), 'myProp');

        expect(myProp).toEqual({ foo: 'bar' });
    });

    it('returns null if a prop is missing', () => {
        document.body.innerHTML = `
            <div id="el"></div>
        `;

        const myProp = prop(query('#el'), 'myProp');

        expect(myProp).toBeNull();
    });

    it('can return a default value if a prop is missing', () => {
        document.body.innerHTML = `
            <div id="el"></div>
        `;

        const myProp = prop(query('#el'), 'myProp', 'default');

        expect(myProp).toBe('default');
    });

    it('throws an error when something goes wrong parsing a JSON prop', () => {
        document.body.innerHTML = `
            <div id="el" :my-prop='{ "foo": "bar }'></div>
        `;

        expect(() => {
            prop(query('#el'), 'myProp');
        }).toThrowErrorMatchingSnapshot();
    });
});

describe('props', () => {

    it('can return an object containing all defined props on an element', () => {
        document.body.innerHTML = `
            <div id="el" prop-one="one" :prop-two="2"></div>
        `;

        const allProps = props(query('#el'));

        expect(allProps).toEqual({ id: 'el', propOne: 'one', propTwo: 2 });
    });
});
