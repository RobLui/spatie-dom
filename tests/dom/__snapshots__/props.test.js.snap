exports[`prop throws an error when something goes wrong parsing a JSON prop 1`] = `"Error parsing JSON for prop my-prop: Unexpected end of JSON input"`;
